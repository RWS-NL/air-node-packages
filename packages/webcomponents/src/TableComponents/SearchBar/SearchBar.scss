@import "../../styles/vars";

.paper {
  align-items: center;
  display: flex;
  height: 2.5rem;
  padding: 2px;
  width: 400px;
}

.input {
  flex: 1;
  font-size: 1rem;
  margin-left: 8px;

  &Button {
    padding: 10px;
  }

  &SVG {
    height: 1.3rem;
    width: 1.3rem;
  }
}
