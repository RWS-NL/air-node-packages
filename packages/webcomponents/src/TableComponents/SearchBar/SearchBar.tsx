/* eslint-disable */
import IconButton from '@material-ui/core/IconButton';
import Paper from '@material-ui/core/Paper';
import SearchIcon from '@material-ui/icons/Search';
import CloseIcon from '@material-ui/icons/Close';
import { CSSProperties } from '@material-ui/styles';
import React, { ChangeEvent, FC, KeyboardEvent, useState, memo } from 'react';
import css from './SearchBar.scss';
import InputBase from '@material-ui/core/InputBase';

export interface SearchBarProps {
  /** Whether to clear search on escape */
  cancelOnEscape?: boolean;
  /** Override or extend the styles applied to the component. */
  classes?: any;
  /** Custom top-level class */
  className?: string;
  /** Disables text field. */
  disabled?: boolean;
  /** Fired when the search is cancelled. */
  onCancelSearch?: () => any;
  /** Fired when the text value changes. */
  onChange?: (event: ChangeEvent<HTMLInputElement>) => any;
  /** Fired when the search icon is clicked. */
  onRequestSearch?: (event: any) => any;
  /** Fired when the search bar is focused */
  onFocus?: (event: any) => any;
  /** Fired when the focus on the search bar is lost */
  onBlur?: (event: any) => any;
  /** Fired when a key is released */
  onKeyUp?: (event: KeyboardEvent<HTMLInputElement>) => any;
  /** Sets placeholder text for the embedded text field. */
  placeholder?: string;
  /** Override the inline-styles of the root element. */
  style?: CSSProperties;
  /** Override the paper element elevation */
  paperElevation?: number;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
};

/**
 * Constructs a search bar using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the searchbar
 * @example
 * <SearchBar
 *   data-qa='table-search-bar'
 *   placeholder='Search...'
 *   onChange={e => debouncedSearch(e.target.value)}
 *   onCancelSearch={console.log}
 *   paperElevation={2}
 * />
 */
export const SearchBar: FC<SearchBarProps> = props => {
  const [value, setValue] = useState('');
  const placeholder = props.placeholder || 'Search...';

  const handleFocus = (event: any): void => {
    if (props.onFocus) props.onFocus(event);
  };

  const handleBlur = (event: any): void => {
    if (value && value.trim().length === 0) setValue('');
    if (props.onBlur) props.onBlur(event);
  };

  const handleInput = (event: ChangeEvent<HTMLInputElement>): void => {
    setValue(event.target.value);
    if (props.onChange) props.onChange(event);
  };

  const handleCancel = (): void => {
    setValue('');
    if (props.onCancelSearch) props.onCancelSearch();
  };

  const handleRequestSearch = (): void => {
    if (props.onRequestSearch) props.onRequestSearch(value);
  };

  const handleKeyUp = (event: KeyboardEvent<HTMLInputElement>): void => {
    if (event.charCode === 13 || event.key === 'Enter') {
      handleRequestSearch();
    } else if (props.cancelOnEscape && (event.charCode === 27 || event.key === 'Escape')) {
      handleCancel();
    }

    if (props.onKeyUp) props.onKeyUp(event);
  };

  return (
    <Paper className={css.paper} elevation={props.paperElevation} square>
      <InputBase
        className={css.input}
        placeholder={placeholder}
        inputProps={{ 'aria-label': placeholder }}
        onBlur={handleBlur}
        value={value}
        onChange={handleInput}
        onKeyUp={handleKeyUp}
        onFocus={handleFocus}
        disabled={props.disabled}
        data-qa={props['data-qa']}
        fullWidth
      />
      {
        !value
          ?
          <IconButton aria-label={placeholder} className={css.inputButton} onClick={handleRequestSearch} disabled={props.disabled} size='small' color='primary' data-qa='search-button'>
            <SearchIcon className={css.inputSVG} />
          </IconButton>
          :
          <IconButton aria-label={placeholder} className={css.inputButton} onClick={handleCancel} disabled={props.disabled} size='small' color='primary' data-qa='search-close-button'>
            <CloseIcon className={css.inputSVG} />
          </IconButton>
      }

    </Paper>
  );
};

export default memo(SearchBar);