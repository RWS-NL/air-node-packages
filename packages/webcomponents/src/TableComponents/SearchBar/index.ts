export { default as SearchBar, SearchBarProps } from './SearchBar';