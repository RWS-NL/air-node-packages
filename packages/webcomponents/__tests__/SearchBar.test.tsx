import { shallow } from 'enzyme';
import React from 'react';
import SearchBar from '../src/SearchBar/SearchBar';

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const searchBar = shallow(<SearchBar />);
    expect(searchBar).toMatchSnapshot();
  });

  test('Optional Props', () => {
    const searchBar = shallow(
      <SearchBar
        cancelOnEscape disabled
        paperElevation={3}
        placeholder='Snaptest'
        data-qa='snap-qa'
        onCancelSearch={jest.fn()}
        onChange={jest.fn()}
        onBlur={jest.fn()}
        onFocus={jest.fn()}
        onKeyUp={jest.fn()}
        onRequestSearch={jest.fn()}
      />
    );
    expect(searchBar).toMatchSnapshot();
  });
});