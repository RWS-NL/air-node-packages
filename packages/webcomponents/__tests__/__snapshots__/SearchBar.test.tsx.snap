// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Optional Props 1`] = `
<WithStyles(ForwardRef(Paper))
  className="paper"
  elevation={3}
  square={true}
>
  <WithStyles(ForwardRef(InputBase))
    className="input"
    data-qa="snap-qa"
    disabled={true}
    fullWidth={true}
    inputProps={
      Object {
        "aria-label": "Snaptest",
      }
    }
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    placeholder="Snaptest"
    value=""
  />
  <WithStyles(ForwardRef(IconButton))
    aria-label="Snaptest"
    className="inputButton"
    color="primary"
    disabled={true}
    onClick={[Function]}
    size="small"
  >
    <SearchIcon
      className="inputSVG"
    />
  </WithStyles(ForwardRef(IconButton))>
</WithStyles(ForwardRef(Paper))>
`;

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Paper))
  className="paper"
  square={true}
>
  <WithStyles(ForwardRef(InputBase))
    className="input"
    fullWidth={true}
    inputProps={
      Object {
        "aria-label": "Search...",
      }
    }
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    placeholder="Search..."
    value=""
  />
  <WithStyles(ForwardRef(IconButton))
    aria-label="Search..."
    className="inputButton"
    color="primary"
    onClick={[Function]}
    size="small"
  >
    <SearchIcon
      className="inputSVG"
    />
  </WithStyles(ForwardRef(IconButton))>
</WithStyles(ForwardRef(Paper))>
`;
